### PR TITLE
bean/ch-274_house_ads

### PR DESCRIPTION
**Pull request by StareIntoTheBeard**
[ch274]

### CODE REVIEW

File: lib/ad_service/query/for_display.ex [hunk 1]
```diff
@@ -2,7 +2,28 @@ defmodule AdService.Query.ForDisplay do
   import Ecto.Query
   alias AdService.Advertisement
   alias CodeFund.Campaigns
-  alias CodeFund.Schema.{Audience, Campaign, Creative, Impression}
+  alias CodeFund.Schema.{Audience, Campaign, Creative, Impression, Property}
+
+  def fallback_ad_by_property_id(property_id) do
```
Comment (coderberry) [inline lib/ad_service/query/for_display.ex]: Does this mean that the fallback campaign doesn't have to have a budget set or even be active?

File: lib/code_fund_web/controllers/api/ad_serve_controller.ex [hunk 2]
```diff
@@ -138,6 +124,65 @@ defmodule CodeFundWeb.API.AdServeController do
...
+      ip: conn.remote_ip |> Tuple.to_list() |> Enum.join("."),
+      browser_height: height,
+      browser_width: width
+    }
+
+    case property_id |> AdService.Query.ForDisplay.fallback_ad_by_property_id() do
+      %Advertisement{} = advertisement ->
+        {:ok, %Impression{id: impression_id}} =
+          params
+          |> Map.merge(%{campaign_id: advertisement.campaign_id, house_ad: true})
```
Comment (coderberry) [inline lib/code_fund_web/controllers/api/ad_serve_controller.ex]: Can we ensure that the revenue_amount and distribution_amounts here are set to zero?

File: lib/code_fund/schemas/audience.ex [hunk 3]
```diff
@@ -1,9 +1,13 @@
 defmodule CodeFund.Schema.Audience do
   use CodeFundWeb, :schema
+  alias CodeFund.Schema.Campaign
 
   @primary_key {:id, :binary_id, autogenerate: true}
   @foreign_key_type :binary_id
   schema "audiences" do
+    belongs_to(:campaigns, Campaign, foreign_key: :fallback_campaign_id, define_field: false)
```
Comment (coderberry) [inline lib/code_fund/schemas/audience.ex]: Should this be plural? `campaigns` vs `campaign` ?

File: test/ad_service/query/for_display_test.exs [hunk 4]
```diff
@@ -83,6 +83,27 @@ defmodule AdService.Query.ForDisplayTest do
...
+      property =
+        insert(:property, audience: insert(:audience, fallback_campaign_id: fallback_campaign.id))
+
+      assert AdService.Query.ForDisplay.fallback_ad_by_property_id(property.id) ==
+               %AdService.Advertisement{
+                 body: "This is a Test Creative",
+                 campaign_id: fallback_campaign.id,
+                 headline: "Creative Headline",
+                 image_url: "http://example.com/some.png",
+                 ecpm: Decimal.new("2.00"),
```
Comment (coderberry) [inline test/ad_service/query/for_display_test.exs]: ecpm on a house ad will always be zero

File: test/ad_service/query/for_display_test.exs [hunk 5]
```diff
@@ -83,6 +83,27 @@ defmodule AdService.Query.ForDisplayTest do
...
+                 campaign_name: "Test Campaign",
+                 small_image_object: nil,
```
Comment (coderberry) [inline test/ad_service/query/for_display_test.exs]: Just curious.. I noticed above you removed the small object. Is this not populated in the test for a reason?

File: test/code_fund_web/controllers/api/ad_serve_controller_test.exs [hunk 6]
```diff
@@ -465,6 +465,52 @@ defmodule CodeFundWeb.API.AdServeControllerTest do
...
+        get(conn, ad_serve_path(conn, :details, property, %{"height" => 800, "width" => 1200}))
+
+      impression = CodeFund.Impressions.list_impressions() |> List.first()
+      assert impression.ip == "12.109.12.14"
+      assert impression.error_code == AdService.ImpressionErrors.fetch_code(:no_possible_ads)
+      assert impression.property_id == property.id
+      assert impression.campaign_id == fallback_campaign.id
+      assert impression.browser_height == 800
+      assert impression.browser_width == 1200
+      assert impression.house_ad == true
```
Comment (coderberry) [inline test/code_fund_web/controllers/api/ad_serve_controller_test.exs]: perhaps add a couple of more assertions on revenue_amount and distribution_amount being 0

File: lib/code_fund_web/controllers/api/ad_serve_controller.ex (hunk 2 above)
Comment (StareIntoTheBeard) [reply lib/code_fund_web/controllers/api/ad_serve_controller.ex]: They are by default

File: test/ad_service/query/for_display_test.exs (hunk 4 above)
Comment (StareIntoTheBeard) [reply test/ad_service/query/for_display_test.exs]: Fair enough, revenues arent set in the impressions in this case and here isn’t anything forcing fall back campaigns to have 0 cpm though.

File: lib/ad_service/query/for_display.ex (hunk 1 above)
Comment (StareIntoTheBeard) [reply lib/ad_service/query/for_display.ex]: Yep.

**Pull request merged**